### DISCUSSION TITLE
1.7 -> tmpfs vs fork or execvp

### DISCUSSION THREAD

**Discussion by nmwael** [gramineproject/gramine#1994] [Q&A] [answered]
If a parent enclave writes to tmpfs and then creates a child with either fork or execvp, does the instance of tmpfs survive or is a new instance?

**Comment by dimakuv** [comment]
Current tmpfs implementation says this: https://github.com/gramineproject/gramine/blob/a20d5b8339b880f79b92c4946d9eb5adfed98b66/libos/src/fs/tmpfs/fs.c#L8-L10

So, the child process/enclave will only get those tmpfs files that are currently being open.

To illustrate with an example, `file1` will be visible in the child, but `file2` will not be visible in the child:
```c
int fd1 = open("tmpfs/file1");
write(fd1, "some text");

int fd2 = open("tmpfs/file2");
write(fd2, "some other text");
close(fd2);

fork();

// child gets only fd1 and its associated file "file1"
```

**Reply by nmwael** [reply]
Ahh, ok. Thanks for the explanation and "workaround"

**Reply by nmwael** [reply]
what happens if  I close fd1 after fork, just to be clear? 

int fd1 = open("tmpfs/file1");
write(fd1, "some text");

int fd2 = open("tmpfs/file2");
write(fd2, "some other text");
close(fd2);

fork();
close(fd1);

**Reply by dimakuv** [reply, marked as answer]
> what happens if I close fd1 after fork, just to be clear?

Nothing bad. The file "file1" with its contents will still be there in the child.
